hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/Socket.java

```diff
@@ -0,0 +1,214 @@
+package com.hazelcast.internal.tpc;
+
+import com.hazelcast.internal.tpc.logging.TpcLogger;
+import com.hazelcast.internal.tpc.logging.TpcLoggerLocator;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hazelcast.internal.tpc.util.Preconditions.checkNotNull;
+
+/**
+ * The socket for TPC engine.
+ */
+public abstract class Socket implements Closeable {
+
+    /**
+     * Allows for objects to be bound to this socket. Useful for the lookup of services and other dependencies.
+     */
+    @SuppressWarnings("checkstyle:VisibilityModifier")
+    public final ConcurrentMap<?, ?> context = new ConcurrentHashMap<>();
+
+    protected final TpcLogger logger = TpcLoggerLocator.getLogger(getClass());
+    protected final AtomicReference<State> state = new AtomicReference<>(State.OPEN);
+    private volatile String closeReason;
+    private volatile Throwable closeCause;
+    private CloseListener closeListener;
+    private Executor closeExecutor;
+
+    /**
+     * Configures the close listener.
+     * <p/>
+     * Can only be configured once.
+     * <p/>
+     * Is threadsafe.
+     * <p/>
+     * If the method is called when the socket already is closed, the close listener
+     * is called.
```

Comment:
nit: I would recommend making paragraphs into unordered list (same applies for the below javadoc as well)